clients/da-vinci-client/src/main/java/com/linkedin/davinci/client/DaVinciRecordTransformer.java

```diff
@@ -29,77 +34,176 @@ public abstract class DaVinciRecordTransformer<K, V, O> {
+   *
+   * @return a {@link Schema} corresponding to the type of {@link K}.
+   */
+  public abstract Schema getKeySchema();
+
+  /**
+   * Returns the schema for the output value used in {@link DaVinciClient}'s operations.
    *
    * @return a {@link Schema} corresponding to the type of {@link O}.
    */
-  public abstract Schema getKeyOutputSchema();
+  public abstract Schema getOutputValueSchema();
 
-  public abstract Schema getValueOutputSchema();
+  /**
+   * Implement this method to transform records before they are stored.
+   * This can be useful for tasks such as filtering out unused fields to save storage space.
+   *
+   * @param key the key of the record to be transformed
+   * @param value the value of the record to be transformed
+   * @return {@link DaVinciRecordTransformerResult}
+   */
+  public abstract DaVinciRecordTransformerResult<O> transform(Lazy<K> key, Lazy<V> value);
 
   /**
-   * @param key to be put
-   * @param value to be put
-   * @return the object to keep in storage, or null if the put should be skipped
+   * Implement this method to manage custom state outside the Da Vinci Client.
+   *
+   * @param key the key of the record to be put
+   * @param value the value of the record to be put,
+   *              derived from the output of {@link #transform(Lazy key, Lazy value)}
    */
-  public abstract O put(Lazy<K> key, Lazy<V> value);
+  public abstract void processPut(Lazy<K> key, Lazy<O> value);
 
   /**
-   * By default, deletes will proceed. This can be overridden if some deleted records should be kept.
+   * Override this method to customize the behavior for record deletions.
+   * For example, you can use this method to delete records from a custom storage outside the Da Vinci Client.
+   * By default, it performs no operation.
    *
-   * @param key to be deleted
-   * @return the object to keep in storage, or null to proceed with the deletion
+   * @param key the key of the record to be deleted
    */
-  public O delete(Lazy<K> key) {
-    return null;
-  }
+  public void processDelete(Lazy<K> key) {
+    return;
+  };
 
   /**
-   * This function is called as a lifecycle event at the beginning of an ingestion task.
+   * Lifecycle event triggered before records are consumed.
+   * Use this method to perform setup operations such as opening database connections or creating tables.
    *
    * By default, it performs no operation.
-   *
    */
-  public void onStartIngestionTask() {
+  public void onStart() {
     return;
   }
 
   /**
-   * This function is called as a lifecycle event at the end of an ingestion task
+   * Lifecycle event triggered when the current {@link #storeVersion} is being retired.
+   * Use this method to perform cleanup operations such as closing database connections or dropping tables.
    *
    * By default, it performs no operation.
-   *
    */
-  public void onEndIngestionTask() {
+  public void onEnd() {
     return;
   }
```

Comment:
Since we're abstracting away the ingestion task from the user, I feel like `onEnd` isn't descriptive enough about what the function does. I'm thinking `onShutdown` would be better.

Reply:
Renaming to `onEndVersionIngestion`